runtime/vm/jvminit.c

```diff
@@ -1818,19 +1819,39 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
 				IDATA argIndexCompactOnIdleDisable = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXIDLETUNINGCOMPACTONIDLEDISABLE, NULL);
 				IDATA argIndexIgnoreUnrecognizedOptionsEnable = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXIDLETUNINGIGNOREUNRECOGNIZEDOPTIONSENABLE, NULL);
 				IDATA argIndexIgnoreUnrecognizedOptionsDisable = FIND_AND_CONSUME_ARG(EXACT_MATCH, VMOPT_XXIDLETUNINGIGNOREUNRECOGNIZEDOPTIONSDISABLE, NULL);
+				BOOLEAN enableGcOnIdle = FALSE;
+				BOOLEAN inContainer = FALSE;
+				
+				omrsysinfo_is_running_in_container(&inContainer);
 
 				/* 
-				 * idle heap tuning is enabled only if -XX:+IdleTuningGcOnIdle is set 
-				 * or if java version is 9 or above and -Xtune:virtualized is set as VM option
+				 * GcOnIdle is enabled only if:
+				 * 1. -XX:+IdleTuningGcOnIdle is set, or
+				 * 2. running in container, or
+				 * 3. if java version is 9 or above and -Xtune:virtualized is set as VM option
 				 */
-				if ((argIndexGcOnIdleEnable > argIndexGcOnIdleDisable) 
-				|| ((J2SE_VERSION(vm) >= J2SE_19) && (argIndexGcOnIdleDisable == -1) && J9_ARE_ANY_BITS_SET(vm->runtimeFlags, J9_RUNTIME_TUNE_VIRTUALIZED))
-				) {
+				if (argIndexGcOnIdleEnable > argIndexGcOnIdleDisable) {
+					enableGcOnIdle = TRUE;
+				} else if (-1 == argIndexGcOnIdleDisable) {
+					if (inContainer
+						|| ((J2SE_VERSION(vm) >= J2SE_19) && J9_ARE_ANY_BITS_SET(vm->runtimeFlags, J9_RUNTIME_TUNE_VIRTUALIZED))
+					) {
+						enableGcOnIdle = TRUE;
+					}
+				}
+				if (TRUE == enableGcOnIdle) {
 					vm->vmRuntimeStateListener.idleTuningFlags |= (UDATA)J9_IDLE_TUNING_GC_ON_IDLE;
 				} else {
 					vm->vmRuntimeStateListener.idleTuningFlags &= ~(UDATA)J9_IDLE_TUNING_GC_ON_IDLE;
 				}
-				if (argIndexCompactOnIdleEnable > argIndexCompactOnIdleDisable) {
+				/*
+				 * CompactOnIdle is enabled only if:
+				 * 1. -XX:+IdleTuningCompactOnIdle is set, or
+				 * 2. running in container
+				 */
```

Comment:
Why different conditions for CompactOnIdle vs IdleTuningGcOnIdle?

Reply:
Only missing condition for CompactOnIdle is presence of -Xtune:virtualized option. I am not sure why we did not add that condition for CompactOnIdle. PR where IdleTuningGcOnIdle was enabled for -Xtune:virtualized is:  https://github.com/eclipse/openj9/pull/125.
But I think that should be considered separately from this PR.

Reply:
@ashu-mehra if it is to be considered separately, there should be an issue for it.

Reply:
@pshipton ok, I will open an issue to address this.

Reply:
Issue to address this concern #4004